Fix unreliable behavior when special vars are readonly or unset

JohnoKing opened the pull request:
This pull request fixes the bug detailed in #4. When a virtual subshell finishes, `$RANDOM` and `$LINENO` are ~unset~ restored with `_nv_unset(np, 0)`:
https://github.com/ksh93/ksh/blob/99065353b38715f1533379d549568630537a5f76/src/cmd/ksh93/sh/init.c#L639-L642


This is incorrect, as it doesn't tell `_nv_unset` to ~unset~ restore the variables if they are readonly. This is fixed by adding `NV_RDONLY` to the `_nv_unset` call, like in other variables that aren't affected by this bug:
https://github.com/ksh93/ksh/blob/99065353b38715f1533379d549568630537a5f76/src/cmd/ksh93/sh/init.c#L781-L782

This pull request also fixes another bug with `${.sh.lineno}`. If `unset .sh.lineno` is run, a memory fault will occur as ksh will try to free it from memory. `${.sh.lineno}` cannot be freed from memory because it is an integer; for reference here are the entries for `${.sh.subshell}` and `${.sh.lineno}` in the default variable table:
https://github.com/ksh93/ksh/blob/99065353b38715f1533379d549568630537a5f76/src/cmd/ksh93/data/variables.c#L96-L98

Fixing the memory fault only requires `${.sh.lineno}` to be given the `NV_NOFREE` flag.

EDIT: Corrected 'unset' to 'restore'.

McDutchie commented on src/cmd/ksh93/sh/subshell.c:
```diff
@@ -760,6 +760,7 @@ Sfio_t *sh_subshell(Shell_t *shp,Shnode_t *t, volatile int flags, int comsub)
 		shp->subshell--;		/* decrease level of virtual subshells */
 		SH_SUBSHELLNOD->nvalue.s--;	/* decrease ${.sh.subshell} */
 	}
+
```
Extra empty line cluttering up the diff.

McDutchie commented on src/cmd/ksh93/tests/variables.sh:
This should be "Check unset and cleanup/restore behavior of special variables". Sorry, I know that's my own comment!

McDutchie commented on src/cmd/ksh93/sh/subshell.c:
```diff
@@ -252,7 +252,7 @@ Namval_t *sh_assignok(register Namval_t *np,int add)
 		return(np);
 	/* don't bother with this */
-	if(!sp->shpwd || np==SH_LEVELNOD || np==L_ARGNOD || np==SH_SUBSCRNOD || np==SH_NAMENOD)
+	if(!sp->shpwd || (add != 2 && (np==SH_LEVELNOD || np==L_ARGNOD || np==SH_SUBSCRNOD || np==SH_NAMENOD)))
```
I would like to extend the comment before the `sh_assignok()` function definition to better explain what it does and what the `add` parameter means (including the new value).

I'm still not sure I get it completely. Does the following look correct to you?

```c
/*
 * This routine will make a copy of the given node in the
 * layer created by the most recent virtual subshell if the
 * node hasn't already been copied.
 *
 * add == 0:    Move the node pointer from the parent shell to the current virtual subshell.
 * add == 1:    Create a copy of the node pointer in the current virtual subshell.
 * add == 2:    For restoring variable state when exiting virtual subshell. Same as 1, but also
 *              disable optimization for ${.sh.level}, $_, ${.sh.subscript} and ${.sh.name}.
 */
```

JohnoKing replied:
The comment is a good idea (added to the pull request), although I have revised it with a different explanation as `sh_assignok(np, 1)` is also meant to save the variable state for when a virtual subshell exits (it just doesn't guarantee it).

McDutchie reviewed (changes requested):
Brilliant, my respect for finding that fix! Nearly done now…

JohnoKing committed:
Fix unreliable behavior when special vars are readonly or unset

src/cmd/ksh93/data/variables.c:
 - Running `unset .sh.lineno` creates a memory fault, so fix that
   by giving it the NV_NOFREE attribute. This crash was happening
   because ${.sh.lineno} is an integer that cannot be freed from
   memory with free(3).

src/cmd/ksh93/sh/init.c:
 - Tell _nv_unset to ignore NV_RDONLY when $RANDOM and $LINENO are
   restored from the subshell scope. This is required to fully
   restore the original state of these variables after a virtual
   subshell finishes.

src/cmd/ksh93/bltins/typeset.c,
src/cmd/ksh93/sh/subshell.c:
 - Disabled some optimizations for two instances of `sh_assignok` to
   fix `readonly` in virtual subshells and `(unset .sh.level)` in
   nested functions. This fixes the following variables when `(readonly
   $varname); enum varname=` is run:

   $_
   ${.sh.name}
   ${.sh.subscript}
   ${.sh.level}

   The optimization in question prevents sh_assignok from saving the
   original state of these variables by making the sh_assignok call
   a no-op. Ksh needs the original state of a variable for it to be
   properly restored after a virtual subshell has run, otherwise ksh
   will simply carry over any new flags (being NV_RDONLY in this case)
   from the subshell into the main shell.

src/cmd/ksh93/tests/variables.sh:
 - Add regression tests from @McDutchie for setting special variables
   as readonly in virtual subshells and for unsetting special
   variables in general.

Fixes #4
